### PR TITLE
Add a new provider - Barry (Denmark)

### PR DESCRIPTION
**Pull request by marval**
Stumbled upon your project and wanted to add my own provider - Barry. I believe I have done it the same way you have added others, but let me know if you find any issues.

The current call is made towards Barry's API for Denmark. I believe they are also a provider in France, but have no idea whether or not they have the same API.

### CODE REVIEW

File: README.md
Comment (MattJeanes) [inline README.md:79]: I assume this line can be removed? I can't see any VAT Multiplier implementation for Barry

File: TeslaMateAgile/Data/Options/BarryOptions.cs
Comment (MattJeanes) [inline TeslaMateAgile/Data/Options/BarryOptions.cs:9]: These two options as they are already in the `BarryOptions` class can just be renamed to `ApiKey` and `MPID` respectively, the Barry prefix is redundant here

File: TeslaMateAgile/Services/BarryService.cs
Comment (MattJeanes) [inline TeslaMateAgile/Services/BarryService.cs:23]: nit: use var here
```suggestion
        var request = new BarryRequest {
```

File: TeslaMateAgile/Services/BarryService.cs
Comment (MattJeanes) [inline TeslaMateAgile/Services/BarryService.cs:32]: The variable name here `agileResponse` I assume was copied from the Octopus provider but doesn't make much sense here, maybe `barryResponse`?

File: TeslaMateAgile/Services/BarryService.cs
Comment (MattJeanes) [inline TeslaMateAgile/Services/BarryService.cs:24]: nit: use lowercase string

```suggestion
            Params = new string[] { _options.BarryMPID, from.UtcDateTime.ToString("o"), to.UtcDateTime.ToString("o")}.ToList()
```

File: TeslaMateAgile/Services/BarryService.cs
Comment (MattJeanes) [inline TeslaMateAgile/Services/BarryService.cs:38]: I believe you shouldn't have to reference BarryService here but keep this in if it is required

```suggestion
            ValidFrom = ParseIso8601(x.Start),
            ValidTo = ParseIso8601(x.End)
```

File: TeslaMateAgile/appsettings.json
Comment (MattJeanes) [inline TeslaMateAgile/appsettings.json:25]: If these are just placeholder values they can be removed here, the appsettings.json is for default values that the user doesn't generally need to touch e.g. the `BaseUrl` as you have added

File: TeslaMateAgile/Data/Options/BarryOptions.cs
Comment (MattJeanes) [inline TeslaMateAgile/Data/Options/BarryOptions.cs:10]: These two look to be required fields so can we add the `[Required]` annotation to these as well? Also split the properties up with empty lines between as they've got annotations now
```suggestion

    [Required]
    public string BarryApiKey { get; set; }
    
    [Required]
    public string BarryMPID { get; set; }
```

**Review by MattJeanes — CHANGES_REQUESTED** [review]
First of all many thanks for contributing! You're the first direct contributor other than myself to the project so welcome!

I can't really test this myself so I'll have to take your word for it but everything looks pretty sensible to me.

I have added some mostly formatting/style comments as I'm a little bit of a perfectionist (sorry)  but it looks great! Once again thank you for contributing 😄 

P.S. What a name for an energy provider!

**Comment by marval** [comment]
.NET is definitely not my strongest programming language :D A lot of it was copy/paste from your other classes. I will take a look at the comments and address them later today. Funny how I am the first contributor :D Having variable pricing makes the charging costs in TeslaMate a nightmare ...

On a side note - we (or at least many users) have reimbursements in Denmark. Basically you can get a reduction of the price per kWh for charging your car if you have a subscription, PVs or heatpump. This is a modification that I am wondering where to place - TeslaMate or here. Would you be open to something so country specific?

**Comment by MattJeanes** [comment]
You've done very well even if it's not your strongest language! Thanks, I sometimes feel I might be a bit too nitpicky but that is just the perfectionist in me, don't take my comments personally 😄 

> On a side note - we (or at least many users) have reimbursements in Denmark. Basically you can get a reduction of the price per kWh for charging your car if you have a subscription, PVs or heatpump. This is a modification that I am wondering where to place - TeslaMate or here. Would you be open to something so country specific?

I can see such a system existing in other countries in the future if not already and even if not I'm happy for country-specific functionality (after all, most of the energy providers are country-specific). It's up to you where you want to make this change, but I would note that you'd probably have to make it here anyway to work with dynamic pricing providers.

If implemented in this project, it may be worth having a global option for it that works across all providers if that makes sense rather than specific to certain providers. Then by using the FixedPrice provider with the same price all day you can even effectively match standard TeslaMate behaviour with the extra goodies provided by this project such as this discount.

Feel free to reach out if you need any help on that one if you do decide to do that in this project!

File: README.md
Comment (MattJeanes) [inline README.md:75]: You'll want to edit the list of supported energy providers / tariffs near the top of this file to include your new one

**Review by marval — COMMENTED** [review]
All should be fixed now

**Comment by marval** [comment]
Seems that reimbursements can be handled by negative FeePerKilowattHour so no need for development there. I am already planning another provider so looking forward to addressing all the issues on Barry so I know what mistakes I shall not repeat.

**Review by MattJeanes — APPROVED** [review]
Those changes look great, thank you! And good to know the `FeePerKilowattHour` works in reverse too - didn't even think of that!

Already looking forward to seeing your next provider!

Just an obvious question before I merge this in and release it, but I assume you've ran this all up and it works? I'm of course unable to test your energy provider especially without any API keys 😄

File: README.md
Comment (MattJeanes) [inline README.md:81]: Oops, did just spot this little documentation error after renaming the variables!

```suggestion
- Barry__ApiKey=XXXXX # Required - API key can be optained through their mobile app
- Barry__MPID=YYYYY # Required - metering point ID (MPID) can be obtained through their mobile app
```

File: TeslaMateAgile/appsettings.json
Comment (MattJeanes) [inline TeslaMateAgile/appsettings.json:23]: Last item in array should not have comma on it

```suggestion
    "BaseUrl": "https://jsonrpc.barry.energy/json-rpc"
```

File: TeslaMateAgile/appsettings.json
Comment (MattJeanes) [inline TeslaMateAgile/appsettings.json:24]: Ditto on comma for this line too:
image


```suggestion
  }
```

File: TeslaMateAgile/Services/BarryService.cs
Comment (MattJeanes) [inline TeslaMateAgile/Services/BarryService.cs:3]: Can simplify serializer name below

```suggestion
using System.Text.Json;
using System.Text.Json.Serialization;
```

File: TeslaMateAgile/Services/BarryService.cs
Comment (MattJeanes) [inline TeslaMateAgile/Services/BarryService.cs:26]: Name can be simplified

```suggestion
        var objAsJson = JsonSerializer.Serialize(request);
```

File: TeslaMateAgile/Services/BarryService.cs
Comment (MattJeanes) [inline TeslaMateAgile/Services/BarryService.cs:32]: Name can be simplified

```suggestion
        var barryResponse = await JsonSerializer.DeserializeAsync<BarryResponse>(await resp.Content.ReadAsStreamAsync());
```

**Review by MattJeanes — CHANGES_REQUESTED** [review]
Sorry just a couple more tidbits my Visual Studio IDE suggested while I was browsing the code! Again just formatting nitpicks 😄

**Review by MattJeanes — APPROVED** [review]

**Comment by MattJeanes** [comment]
Looking great on the code now @marval - sorry for all the nitpicking!

I'm assuming this is all working good for you? I'm unable to test it without an API key unless they have a sample one somewhere with test data but I couldn't find one on their API documentation. Happy to merge + release if it's all working well for you 😄

**Comment by marval** [comment]
There is a bit of an issue for the users that have a lot of charges to calculate price for - they are reaching Barry's short term throttling (10 requests per minute), but they would get 10 charges for each run. I might just add a bit about  this in the readme just to clarify it for everyone.

**Comment by MattJeanes** [comment]
Yeah that might be good to note, but if it will eventually go through on the subsequent retries and doesn't get totally stuck I think that's fine.

Might be a good idea to add a new section to the 'Energy provider setup' in the README and move all information on setup into there like how to get the MPID / API key including this note about rate limiting, then in the YAML bit you can change the comment to something like `# See below Barry section`. You can see a similar use of this with the Tibber provider.

**Comment by marval** [comment]
I added the section as suggested. Hope to get some more users to test and I am ready to fix their issues if they occur.

**Comment by gahms** [comment]
I just wanted to let you know that I am now using TeslaMateAgile with Barry integration and made a few sanity checks by hand and it seems to work perfectly!

**Comment by MattJeanes** [comment]
That's great @gahms - cheers for letting me know!

**Comment by PalleRaa** [comment]
Just found this thread now. It is also working for me. No problem. Only issue now is that Barry has announced that they are closing :(
I have been trying to persuade @marval to add the Nordpool prices to TeslaMateAgile. A lot of people in DK, SE, NO, FI, DE, FR, NL etc gets their prices from this place. Only issue is that Nordpool only shows the raw electricity price. Every vendor will then add different traifs, transportation cost, VAT etc. This would need to be handled by TeslaMateAgile. There exists already a powerfull integration to HomeAssistant that allows you to customize all these extra costs. Maybe that can be used to inspiration https://github.com/custom-components/nordpool

I have very little programming knowledge myself, but will be happy to test and report bugs :)

**Comment by gahms** [comment]
Yes I got the email also:( I already thought about implementing what you talk about. I will have a look at HomeAssistant component you link to and see if I can figure out how to use it for a TeslaMateAgile integration.

**Comment by MattJeanes** [comment]
Looks like a Python implementation of Nordpool prices is here https://github.com/custom-components/nordpool/blob/master/custom_components/nordpool/aio_price.py

I'm always happy for people to contribute and if there's high enough demand and I have access to the datasource myself which I should do with this then I could try adding it in if no-one else wants to give it a go.

Sad to hear Barry is shutting down, hopefully this can take it's place soon.

**Comment by marval** [comment]
Hi all,
Yes ... unfortunately barry is about to close in 3 weeks :( Sad stuff ... I will be looking into getting prices from Nordpool but I am assuming I might have to combine it with the already existing fixed price module to handle the taxes (afgifter) for distribution and so on. If one of you starts on this please let me know so we can combine our efforts - @gahms and @PalleRaa

**Pull request merged**
